### PR TITLE
daprd wait app ready at starting phase

### PR DESCRIPTION
**Pull request by theonefx**
# Description

This PR provides an optional capability for daprd to check and wait for the app to complete startup, this ensure that the app is ready to handle incoming traffic before daprd starts receiving it.

there are two optional check point that developer can use:
- url
- k8s containers name

## Issue reference

this is a part of issue : https://github.com/dapr/dapr/issues/2918

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [X] Unit tests passing
* [X] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/runtime/runtime.go [hunk 1]
```diff
@@ -1910,6 +1920,60 @@ func (a *DaprRuntime) blockUntilAppIsReady() {
...
+
+		if err == nil && resp.Status().Code == nethttp.StatusOK {
+			log.Infof("application discovered on readiness address %s", a.runtimeConfig.ReadinessAddress)
+			break
+		}
+
+		time.Sleep(time.Millisecond * 200)
+	}
+
+	log.Infof("wait applicatoin ready, continue init")
```
Comment (jjcollinge) [inline pkg/runtime/runtime.go:1974]: nit: typo

File: tests/config/dapr_kafka_bindings_custom_route.yaml [hunk 2]
```diff
@@ -12,19 +12,19 @@ spec:
...
-    value: group2
-  # publisher configuration: topic
-  - name: publishTopic
-    value: dapr-test
-  - name: authRequired
-    value: "false"
-    # set specify route
-  - name: route
-    value: /custom-path
+    # Kafka broker connection setting
```
Comment (daixiang0) [inline tests/config/dapr_kafka_bindings_custom_route.yaml:15]: seems no truly changes here.

Comment (theonefx) [reply tests/config/dapr_kafka_bindings_custom_route.yaml:15]: you are rignt, i will revert it.

Comment (theonefx) [reply tests/config/dapr_kafka_bindings_custom_route.yaml]: revert done

File: pkg/runtime/runtime.go (hunk 1 above)
Comment (theonefx) [reply pkg/runtime/runtime.go]: it's a spelling mistake. fixed ^_^

**Review by pkedy — COMMENTED** [review]
This is an interesting approach! I see a new resource permission is required in the ClusterRoleBinding. Hypothetically, how does the code handle not having this permission?

**Comment by theonefx** [comment]
> This is an interesting approach! I see a new resource permission is required in the ClusterRoleBinding. Hypothetically, how does the code handle not having this permission?

I've tried, and I got an error message with no permissions. So, I have to add the permission definition.

File: pkg/runtime/wait.go [hunk 3]
```diff
@@ -69,3 +75,108 @@ func checkIfOutboundReady(client *http.Client, outboundReadyHealthURL string) er
...
+					}
+				}
+			}
+
+			log.Infof("all containers are ready")
+			break
+		}
+	}
+
+	if a.runtimeConfig.WaitingProbe != "" {
```
Comment (artursouza) [inline pkg/runtime/wait.go:148]: Doesn't K8s already have readiness check? I see that this is helpful for Standalone mode but this is too specific - it is on HTTP only (does not support gRPC) and content-type is forced to JSON. Can the waitingProbe (maybe readinessCheck is a better name here) be configured in the Configuration CRD?

Again, this check only adds value to standalone mode.

**Review by artursouza — CHANGES_REQUESTED** [review]
/cc @yaron2

File: pkg/injector/pod_patch.go [hunk 4]
```diff
@@ -593,6 +593,14 @@ func getSidecarContainer(annotations map[string]string, id, daprSidecarImage, im
 				Name:  "NAMESPACE",
 				Value: namespace,
 			},
+			{
+				Name: "POD_NAME",
```
Comment (yaron2) [inline pkg/injector/pod_patch.go:597]: We need to find another way to either pass the pod name to the container or identify the pod in the operator. This uses the FieldPath ref (Downward API) which is not supported on platforms like Virtual Kubelet.

**Review by yaron2 — CHANGES_REQUESTED** [review]

**Comment by yaron2** [comment]
Added comment, looks good otherwise.

**Comment by daixiang0** [comment]
@theonefx please resolve conflicts.

**Comment by yaron2** [comment]
Ping @theonefx

**Comment by daixiang0** [comment]
ping @theonefx

**Pull request closed without merging**